cpp_linter/run.py

```diff
@@ -214,17 +207,28 @@ def verify_files_are_present() -> None:
         repository. If files are not found, then they are downloaded to the working
         directory. This is bad for files with the same name from different folders.
     """
+    files_exist = True
     for file in Globals.FILES:
-        file_name = Path(file["filename"])
+        file_name = Path(file.name)
         if not file_name.exists():
-            logger.warning("Could not find %s! Did you checkout the repo?", file_name)
-            raw_url = f"https://github.com/{GITHUB_REPOSITORY}/raw/{GITHUB_SHA}/"
-            raw_url += urllib.parse.quote(file["filename"], safe="")
-            logger.info("Downloading file from url: %s", raw_url)
-            Globals.response_buffer = requests.get(raw_url)
-            # retain the repo's original structure
-            Path.mkdir(file_name.parent, parents=True, exist_ok=True)
-            file_name.write_text(Globals.response_buffer.text, encoding="utf-8")
+            files_exist = False
+            break
+    if not files_exist:
+        try:
+            repo = pygit2.clone_repository(
+                f"https://github.com/{GITHUB_REPOSITORY}", "."
+            )
+        except ValueError as exc:  # pragma: no cover
+            raise ValueError(
+                "Could not clone repo into working directory because it is not empty"
+            ) from exc
+        commit = repo.revparse_single(GITHUB_SHA)
+        repo.checkout_tree(
+            cast(pygit2.Commit, commit).tree,
+            # reset index to specified commit
+            strategy=pygit2.GIT_CHECKOUT_FORCE | pygit2.GIT_CHECKOUT_RECREATE_MISSING,
+        )
+        repo.set_head(commit.oid)
```

Comment:
This is the only change I'm concerned with. If the user does not checkout the repo but puts files into the `github.workspace` folder before running cpp-linter, then this change will crash the code.

I like this idea because
1. It makes us less dependent on the Github REST API
2. Instead of downloading only changed files, we get the entire snapshot of the repo. This should make clang-tidy happier when unchanged files are still present.
3. It encourages the user to checkout the repo beforehand if the code crashes here.

If this change is not reverted, then I'd like to figure out how to clone a shallow copy at the specified commit. Currently, this doesn't clone a shallow copy.

Reply:
This code failed to communicate to server in CI, thus you had to retry the failed job. Due to this fact, I'm cherry-picking this change to another branch. Maybe we can revisit this idea when I explore gitlab support.